### PR TITLE
Adding notes on nulls/dates for iOS

### PR DESCRIPTION
**Pull request by ikeith**
This branch adds a new 'Data Types' document to the core-apis folder with notes on null and date objects for iOS.

### CODE REVIEW

File: pages/docs/v3/core-apis/data-types.md [hunk 1]
```diff
@@ -0,0 +1,116 @@
+---
+title: Capacitor data types
```
Comment (dwieeb) [inline pages/docs/v3/core-apis/data-types.md:2]:
```suggestion
title: Capacitor Data Types
```

File: package-lock.json [hunk 2]
```diff
@@ -1,8 +1,7056 @@
 {
   "name": "capacitor-site",
   "version": "0.0.1",
-  "lockfileVersion": 1,
+  "lockfileVersion": 2,
```
Comment (dwieeb) [inline package-lock.json:4]: I would undo these changes. (`git rm package-lock.json`)

It looks like you're using npm 7 and most of the team is still on npm 6.

File: pages/docs/v3/core-apis/data-types.md [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+description: Data types in Capacitor
+---
+
+# Capacitor data types
```
Comment (dwieeb) [inline pages/docs/v3/core-apis/data-types.md:6]:
```suggestion
# Capacitor Data Types
```

File: pages/docs/v3/core-apis/data-types.md [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+}
+if let value = value as? Int {
+    // false, the NSNull object cannot be cast to Int
+}
+if let value = value as? NSNull {
+    // true, 'value' is a NSNull
+}
+```
+
+In general, it is not recommended to rely on the presence of a key to convey meaning. Only the corresponding value should be evaluated after its expected type is confirmed. Since dictionaries typically require the typing of each value as it is extracted, the inclusion of `NSNull` can mostly be ignored.
```
Comment (dwieeb) [inline pages/docs/v3/core-apis/data-types.md:49]: It seems like this part is mostly trying to say "don't use `if let value = value` without a cast", wondering if it could maybe be made more concise?

File: pages/docs/v3/core-apis/data-types.md [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+#### Arrays
+
+Since accessing arrays typically requiring typing the whole collection, it is more important to consider if they are heterogeneous or not.
```
Comment (dwieeb) [inline pages/docs/v3/core-apis/data-types.md:53]: 1. 
```suggestion
Since accessing arrays typically require typing the whole collection, it is more important to consider if they are heterogeneous or not.
```

2. I would also consider defining "heterogeneous", or using simpler language.

File: pages/docs/v3/core-apis/data-types.md [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+}
+if let values = call.getArray("bar", Int?) {
+    // false, the array is a mix of Int and NSNull objects and can't be cast to [Int?]
+}
+```
+
+To help with this behavior, Capacitor includes a convenience extension that can map a heterogeneous array with `NSNull` values into an array of optionals. It works on the `JSValue` protocol, which represents all of the valid types that can be bridged between environments, but can be cast to a specific subtype.
+
+```swift
+if let values = call.getArray("bar").capacitor.replacingNullValues() as? [Int?] {
```
Comment (dwieeb) [inline pages/docs/v3/core-apis/data-types.md:70]: Would it make sense if our `getArray` helper did this by default? Or maybe it can't because of Obj-C?

File: pages/docs/v3/core-apis/data-types.md [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+    // true, 'values' is now cast to [Int?] with 'nil' at index 2
+}
+```
+
+---
+
+### Dates
+
+Data moving from the web runtime to native iOS code uses a different mechanism than data going in the other direction. In practice, this difference doesn't matter except for one data type: dates. The WebView can automatically translate JavaScript `Date` objects into native `NSDate` or `Date` objects but dates being returned from a plugin must be serialized as a string.
```
Comment (dwieeb) [inline pages/docs/v3/core-apis/data-types.md:79]: This peculiarity makes me think we should consider switching to JSON serialization/deserialization for iOS. I think plugin developers are expecting it, and any automatic conversion of dates may be unexpected or at least not often used. It would bring parity between the messages and data types between iOS and Android (and other platforms) and I don't think we'll suffer that much of a performance hit (but would be worth benchmarking).

**Review by dwieeb — COMMENTED** [review]
Really well written. 🙌

File: pages/docs/v3/core-apis/data-types.md (hunk 6 above)
Comment (ikeith) [reply pages/docs/v3/core-apis/data-types.md:70]: `getArray` is Swift only so Obj-C is not a concern. The issue is that if we did that by default, then all arrays would get typed as optional regardless of whether or not they contain nulls. I'm trying to avoid optionals where possible to reduce the temptation of force unwrapping (which is why the accessors with a default value no longer return optionals). So it seemed better to put the burden on the less common use-case rather than make everyone cast/test the result.

File: pages/docs/v3/core-apis/data-types.md (hunk 7 above)
Comment (ikeith) [reply pages/docs/v3/core-apis/data-types.md:79]: The problem is that we cannot control how the webview passes the data and it's already handling the conversion. So we could:
1. Reject the payload if it contains dates. But it's not clear how best to message that back to the web view.
2. Serialize/deserialize the dates which doesn't really change anything.
3. Serialize the dates into strings before passing them to the plugin but then we're just doing a different magic conversion, one that might introduce side effects.

If anything, what we'd probably want to do is automatically handle serializing dates into strings when passing data back to the web but it would not be trivial to handle deserializing those back into JavaScript Dates. So we'd still have a mismatch where data transit is different in different directions.

File: pages/docs/v3/core-apis/data-types.md (hunk 4 above)
Comment (ikeith) [reply pages/docs/v3/core-apis/data-types.md]: Yes, that's mostly what it is saying. I've tried to tighten it up.

File: pages/docs/v3/core-apis/data-types.md (hunk 5 above)
Comment (ikeith) [reply pages/docs/v3/core-apis/data-types.md]: Fixed. I've tried for a contextual definition of heterogeneous.

File: pages/docs/v3/core-apis/data-types.md [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+{ 'foo': null, 'bar': [1, 2, null, 4]}
+```
+
+#### Dictionaries
+
+`CAPPluginCall` stores this data as its `options` property but has a variety of convenience accessors that operate on it. The accessors will cast the value to the expected type(s) so `NSNull` values will get filtered out.
+
+```swift
+if let value = call.getString("foo") {
+    // false, `value` is nil
```
Comment (dwieeb) [inline pages/docs/v3/core-apis/data-types.md:30]:
```suggestion
    // GOOD: `value` is nil, so this block won't run
```

File: pages/docs/v3/core-apis/data-types.md [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+}
+```
+
+However, accessing the storage property directly can return an `NSNull` object.
+
+```swift
+if call.options["foo"] != nil {
+    // true, the key returned a value
```
Comment (dwieeb) [inline pages/docs/v3/core-apis/data-types.md:38]:
```suggestion
    // BAD: the key returned a truthy `NSNull` object, so this block will run
```

File: pages/docs/v3/core-apis/data-types.md [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+}
+if let value = call.options["foo"] {
+    // true, 'value' is a valid NSNull object
```
Comment (dwieeb) [inline pages/docs/v3/core-apis/data-types.md:41]:
```suggestion
    // BAD: `value` is a truthy NSNull object, so this block will run
```

File: pages/docs/v3/core-apis/data-types.md [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+}
+```
+
+> It is not recommended to rely on the presence of a key to convey meaning. Always type-check the corresponding value to evaluate it.
+
+#### Arrays
+
+Since accessing an array typically requires typing the whole collection, it is important to consider if it contains a single type or might be heterogenous.
```
Comment (dwieeb) [inline pages/docs/v3/core-apis/data-types.md:49]:
```suggestion
Since accessing an array typically requires typing the whole collection, it is important to consider if it contains a single type or might be heterogeneous.
```

File: pages/docs/v3/core-apis/data-types.md (hunk 7 above)
Comment (dwieeb) [reply pages/docs/v3/core-apis/data-types.md:75]: > we cannot control how the webview passes the data 

Sure we can--all Capacitor calls go through `postToNative`: https://github.com/ionic-team/capacitor/blob/main/core/src/bridge.ts#L37-L44

For Android we run `JSON.stringify()`: https://github.com/ionic-team/capacitor/blob/main/core/src/bridge.ts#L30

File: pages/docs/v3/core-apis/data-types.md (hunk 8 above)
Comment (ikeith) [reply pages/docs/v3/core-apis/data-types.md:30]: I like this pattern of labeling as it's clearer than just true/false. I'm updating all the snippets to match.

File: pages/docs/v3/core-apis/data-types.md (hunk 7 above)
Comment (ikeith) [reply pages/docs/v3/core-apis/data-types.md:72]: @dwieeb After discussing this issue with @theproducer, another possibility would be to handle this natively. That is, we examine the call payload and turn any `NSDate` objects into ISO 8901 strings at the time that [we coerce the types](https://github.com/ionic-team/capacitor/blob/9ff9d1b07e6d6f44781e580183475af9bd8e2e9b/ios/Capacitor/Capacitor/CapacitorBridge.swift#L389).

Reasons
1. Consistency since dates will automatically get turned into strings on both platforms
2. The NSDate-object behavior can be preserved for any Cordova plugins that rely on it (since we've identified the plugin type at that point)
3. The performance impact should be less than stringifying the whole blob in JS first

We could also do the same when serializing data back to JS. So Dates would automatically get stringified in both directions rather than requiring that developers do it (although they'd be free to pre-stringify the date if they wanted some other format).

Comment (dwieeb) [reply pages/docs/v3/core-apis/data-types.md:72]: Sounds good to me.
